lib/src/shared/main/java/com/couchbase/lite/internal/AbstractExecutionService.java

```diff
@@ -112,136 +194,182 @@ public boolean stop(long timeout, @NonNull TimeUnit unit) {
             return false;
         }
 
-        private void finishTask() {
+        void finishTask() {
             final CountDownLatch latch;
             synchronized (this) {
-                running--;
+                if (--running > 0) { return; }
                 latch = stopLatch;
             }
 
             if (latch != null) { latch.countDown(); }
         }
 
-        @VisibleForTesting
-        private void restart() {
+        // Called on completion of the task at the head of the pending queue.
+        void scheduleNext() {
             synchronized (this) {
-                if (stopLatch == null || running > 0) {
-                    throw new IllegalStateException("Executor hasn't been stopped yet.");
+                // the executor has been stopped
+                if (pendingTasks.size() <= 0) { return; }
+
+                do {
+                    pendingTasks.remove();
+                    final InstrumentedTask task = pendingTasks.peek();
+                    if (task == null) { return; }
+                    executeTask(task);
                 }
+                while (spaceAvailable());
```

Comment:
What happens in the case that multiple tasks pile up into the pending task queue, then the first one executes and that pushed it back out of space again.  It seems like this method will exit without completing them all.  When does it come back around when space is available again?

Reply:
Creating two types (one will call finishTask() and the other will call scheduleNext() ) of tasks depending on the available space at the time could be a concern here that that the scheduleNext() might not get called. Maybe when each single task is done, go through the same code path that also tries to schedule the next task?

Reply:
I'm not sure what "pushed out into space again" means.  The currently running task is always the task at the head of the queue.  It can't get lost until it is completed.
I don't see how it is possible that `scheduleNext` could not be called...  Say more?